src/Store.ts

```diff
@@ -36,8 +37,11 @@ export class Store<S> {
 
   dispatch<T extends Redux.Action>(action: T): T {
     this._changeId++;
-    
-    return this.store.dispatch(action);
+    if(isPromise(action))
```

Comment:
Can these be formatted slightly differently? Also, with the defined types we can cast these correctly:

```typescript
if (isPromise<T>(action)) {
  return action.then(this.dispatch.bind(this);
}

if (isFunction(action)) {
  return (<DispatchHandler<T>>action)(this.dispatch.bind(this));
}
```